### DISCUSSION TITLE
Initial discussion

### DISCUSSION THREAD

**Discussion by djdv** [djdv/go-filesystem-utils#5] [General]
In response to: https://github.com/ipfs/go-ipfs/issues/5504#issuecomment-841872458

>Unfortunately that didn't succeed
>Attempting to bind to host system: /fuse/keyfs, /fuse//mnt/keyfs...

Looks like I messed up the path translations somewhere. 
I recently tested and ported over the code that handles the `cmds` pkg stuff, and remember encountering this as well in my Linux environments. (I primarily develop on Windows so Unix multiaddrs were getting messed with a lot in the old versions. Some of the code became redundant after a patch was pulled in `go-multiaddr/net` as well.)

The standalone binary is currently pending reviews here: https://github.com/djdv/go-filesystem-utils/pulls
And there's some dependency issues that need to be patched as well before the CI will clear on Linux.
(I'll be filling issues in the repo later)

But after those are finished, the next section of code to be brought over and tested will be the `mount` command. Specifically meaning; Read-only IPFS and IPNS via Fuse. With the rest of read-only things to follow. Then write support afterwards.

***

>I don't understand why there's only "local" when that is not the name of my "self" key. 

This is how the standard `ipfs mount` implementation handles mounting IPNS. But that's the existing code.
IIRC "local" is just a hardcoded name, and the system uses the MFS api to handle writes and publishes.
There may be some issue in the mount code, or the `mfs` pkg.

In the new implementation, I do a local-only publish to the node every write, and intend to add a full publish after an idle period.
Even if writes are continuous forever and prevent the mount doing a full publish. The `daemon`/node itself should eventually decide to republish its keys (irrelevant of `mount`, but with the new local value).

It sounds similar to how you described but the new implementation still needs to be refactored and stabilized.
I'll have to make sure to look into making tests that mock IPNS and use the file system, then validate publishes are happening when they should be. *At least to some basic degree for now.

Some other differences are that the new implementation distinguishes "ipns" and "keyfs" in its arguments.
So "keyfs" will contain a root that lists the keys as their names, the same as `ipfs key list -l` would; `self` included.
And handles keys that point to supported types (unixfs directories, files, links ...).
The existing implementation handles unixfs directories only IIRC.
The new implementation for mounting "ipns" specifically, has an empty root. But I might change that since it's inconsistent with how `ipfs mount` handles it.

***

>And maybe version control could be done by policy: the config file could say how many old versions to keep, or for how long, so those previous hashes just stay pinned for a while longer

Version control sounds interesting in this way.
I'm sure since it's mountable, you could use conventional file system backup methods to handle it.
Like rsync, etc. but [Venti](https://en.wikipedia.org/wiki/Venti) also comes to mind here.

In regards to the config policies; I think it's viable to handle such policies, and associate them with keys/whatever.
The implementation that comes to mind is just automatically creating some kind of subdirectory within the key, like `/.ipns/versions` similar to how ZFS exposes `/.zfs/snapshot`. And scheduling the tasks / adjusting sync times / using a name other than `.ipns`, etc. in the arguments.
For reference, I've put some consideration into how setting values are being retrieved, and have notes on a `CmdsParameterSet` set [here](https://github.com/djdv/go-filesystem-utils/pull/2) with ideas on extending it later in a way that shouldn't cause too much hassle. It's a bit verbose, but I don't think the declarations can be avoided either way.

>I'm really tired of waiting for ipfs to actually be a filesystem. 

Same. Doing what I can to push it forward a bit. I'd really like to be able to start using the system more seriously/frequently. With things running on top of it as well. Or sourced from it at least (Without exhausting machine resources).

>Theoretically maybe I could help you with this code too. But I have barely written any go at all (I spend most of my time in C++ unfortunately

No worries, the conversation about it alone is super helpful!
But also, I intend for code from any language to be able to interface with the file system service in some ways.

As-is, `fs service --api=...` [is implemented](https://github.com/djdv/go-filesystem-utils/pull/3), and exposes an API in the same way `ipfs daemon --api=...` does. (over whatever multiaddr socket the `cmds` pkg supports, TCP, UDS, etc.)
So C++ should be able to interface with it the same way, but with our own set of commands.
The rest of the program in itself is simply going to be a client of its own service. An implementation `fs list` is [here](https://github.com/djdv/go-filesystem-utils/pull/4) for reference. But is probably going to change as `fs mount` is refactored.

As an aside, my own distant plan is to adapt a [very old Windows shell extensions](https://youtu.be/K6I6fzO1cgQ) I wrote.
Making the user shell more IPFS aware. Allowing for things like "copy cid" in the context menu when viewed from explorer.  ​(Will have to modernize this in C++, like it or not due to the OS API)
![licecap-trim](https://user-images.githubusercontent.com/13862850/119813306-1a08a980-beb7-11eb-8d56-1f3738c57865.gif)

I found this extremely useful for linking files to people quickly, and would like to get that functionality back, but with the modern APIs.
The plan is to have the explorer's context call "/api/list" with encoding like JSON, and just use results.
You can imagine seeing "IPFS->publish" in the context menu while (only) inside an IPNS mountpoint. "IPFS>Unpin", etc.
By communicating with both the `fs` and `ipfs` daemons, in some external client written in another language.

I'm curious what other people might do with the API, and could always extend it if necessary.

>Anyway I look forward to your next push. Thanks for taking this on.

Regards :^)
I appreciate it. Thanks for trying it out and thinking things through with me.

**Comment by polkovnikov** [comment]
@djdv

I have a question a bit unrelated to your project:

Do you know if it is possible somehow to add a huge file in my regular file system to IPFS in such a way that IPFS doesn't duplicate file's content? So if I have 82 GB local file then IPFS doesn't copy 82 GB over into its block-system.

I can imagine that instead of copying whole file IPFS can slice this file into 256 KB blocks and just store light references to these blocks, like file_path + start/end_byte_number. Is there such feature in IPFS?

If there is no such feature in IPFS itself maybe at least your utils are capable of doing such thing?

The main idea for me is to share large data on my local file system without doing duplicating.

Of cause I can use your filesystem util to map IPFS files into regular file system, but I prefer to keep my valuable and sensitive data inside regular file system as a main storage.

Also there is another reason not to move everything into IPFS - because some of my large files I just want to share temporarily, meaning that I don't need them in IPFS at all, I want to share them with my friends for a while, after that it is not needed in IPFS.

**Reply by djdv** [reply]
Hey @polkovnikov 
There is such a feature, however it is experimental and I think it is discouraged as it's sometimes confusing and limited.
It's the `--nocopy` parameter for `add` https://docs.ipfs.io/reference/cli/#ipfs-add
https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#road-to-being-a-real-feature-2

I don't think anyone has made efforts to improve it to the point where it can be considered stable / not experimental. Unfortunately.
You need to do weird hacks like symlinking files under your `$IPFS_PATH`, and I don't remember the details but there's issues around updating the references when files change.
It's probably discussed in here: https://github.com/ipfs/go-ipfs/issues/4224
I saw my own post in there referencing a shell extensions that automates the linking but the `nocopy` feature itself wasn't reliable enough for me to continue using it. Instead I opted to implement mount and migrate my data into IPFS where possible, or accept duplicate copies (for now).

But it would be nice if someone improved that feature, and maybe the same functionality I had in the shell extension could be added to the IPFS desktop project when `nocopy` is not longer experimental.

**Reply by polkovnikov** [reply]
Thanks a lot @djdv , your suggestion worked for me, did following on my Windows:

1) Enabled experimental filestore and urlstore by `ipfs config --json Experimental.FilestoreEnabled true` and `ipfs config --json Experimental.UrlstoreEnabled true`.
2) Moved my `.ipfs` folder from `c:\users\user\.ipfs\` to `d:\misc\.ipfs\` (my huge files are located on disk `d:`).
3) Made directory junction from `d:\misc\.ipfs\` to `c:\users\user\.ipfs\` (through Alt-F6 in [Far File Manager](https://farmanager.com/)). This junction is not really necessary if you set `IPFS_PATH` environment variable instead.
4) Created folder for files links `d:\misc\.ipfs\_linked\`.
5) Made hard link from huge file `d:\data\file.mp4` to `d:\misc\.ipfs\_linked\file.mp4` (also through Alt-F6 in `Far Manager`).
6) Inside folder `c:\users\user\.ipfs\_linked\` did `ipfs add --nocopy file.mp4`. This successfully added my huge file without occupying almost any space inside IPFS block storage. For some reason recent Windows reported error when doing hard links of files between different drives, although previous versions of windows did it successfully, maybe new NTFS (system drive `c:`) has some incompatibilities with older one (data drive `d:`), that's why I had to do all the stuff of moving folder around.
7) File successfully appeared and became readable in your `pinfs` mounted disk.
8) Also tried another command for no-copy adding of remote internet's URL. Did `ipfs add --nocopy https://some.domain.com/huge_file.mp4`, it also added huge file without occupying space in IPFS block storage (200KB actually was occupied).
9) This url-added huge file also appeared in your `pinfs` disk and was readable at any random location, of cause with small delay, because each read does remote HTTPS request.

Both `--nocopy` command increased IPFS block storage occupied space by around 200-500 KB, very tiny, of cause not GBs (sizes of original files). Also it took a bit of time to add local file, seems that IPFS does hashing of all blocks when doing `--nocopy` adding. Url-add showed a progress bar and downloaded whole file, also probably for rehashing purpose, but actual file wasn't stored on my disk.

**Comment by polkovnikov** [comment]
@djdv As you're very familiar (unlike me) with IPFS can you please also suggest is it possible somehow to use IPFS as a proxy of torrents technology?

Meaning if there is any proxy that connects two worlds - torrents and IPFS? They both seem to be very related technologies.

For example if I have `.torrent` file maybe there could be something like `ipfs add --nocopy --torrent huge_file.torrent` which adds content referenced by .torrent file without making a download.

Then when somebody reads a small part of torrented file from IPFS (for example through your mounted disk magic) some underlying code of IPFS node tries to download this small portion from Torrent network.

It looks like a very interesting feature because IPFS and Torrent technologies look very similar to each other. Hence I could imagine that there were some efforts of connecting them two.

**Reply by djdv** [reply]
@polkovnikov 
Hey sorry for the late reply.

I know that this has been talked about before and I assume will be possible eventually, but I don't know the current status of it.
Some relevant links I found searching, there may be other references.
https://github.com/ipfs/js-ipfs/issues/779
https://github.com/mattsse/libp2p-bittorrent-protocol
